cid.go

```diff
@@ -680,3 +680,142 @@ func CidFromBytes(data []byte) (int, Cid, error) {
 
 	return l, Cid{string(data[0:l])}, nil
 }
+
+func toBufByteReader(r io.Reader, dst []byte) *bufByteReader {
+	// If the reader already implements ByteReader, use it directly.
+	// Otherwise, use a fallback that does 1-byte Reads.
+	if br, ok := r.(io.ByteReader); ok {
+		return &bufByteReader{direct: br, dst: dst, alreadyRead: len(dst)}
+	}
+	return &bufByteReader{fallback: r, dst: dst, alreadyRead: len(dst)}
+}
+
+type bufByteReader struct {
+	direct   io.ByteReader
+	fallback io.Reader
+
+	alreadyRead int
```

Comment:
Suggestion:

I got confused by the naming of `alreadyRead` - I thought it mean "the number of bytes that have already been read from the `bufByteReader`", when actually it means "the number of bytes that were already read from the original reader and passed to `bufByteReader`".

I think for me it would be easier to follow the code if instead we have a variable called `index` or similar that counts up from zero, rather than a variable that counts down from the length of `dst`.

Reply:
I thought about this for a while too, but I couldn't figure out a nice enough system. Having two separate byte slices might be clearer for the docs, but it made the code more complex, for example.

As for an index - you mean an index that starts at 0, reuses bytes up to len(dst), and then does reads after that?

Reply:
> you mean an index that starts at 0, reuses bytes up to len(dst), and then does reads after that?

Yes exactly - IIUC that's what the code is doing, but in reverse. In my mind it would be simpler to count up than count down.

Reply:
Sounds good - done. I ended up calling it "consumed", as that felt a bit more descriptive than "index". It also made the constructor a bit simpler.